src/mpi/init/initthread.c

```diff
@@ -221,16 +221,22 @@ int MPIR_Init_thread(int *argc, char ***argv, int required, int *provided)
     int has_args;
     int has_env;
     int thread_provided = 0;
-    int exit_init_cs_on_failure = 0;
```

Comment:
We initialize and enter the `ALLFUNC` mutex first thing so we'll always need to destroy it on failure. We initialize the other mutexes last thing after any possible failures.